### ISSUE TITLE
Why no recoverable errors?

### ISSUE THREAD

**Issue by TerranWorks** [petgraph/petgraph#654]
Just a question, but I can't add the question tag. I'm just getting into petgraph because I know it to be the standard graph rust graph library, but I'm surprised that every operation panics instead of returning a result. I understand that panics aren't evil, but most other crates that I've worked with (Bevy for example) provide sister methods for those that panic which return results instead: e.g. graph.add_node(...) panics, so there would also be a graph.try_add_node(...) which returns a result that can be handled. Is there a reason why results don't seem to be used in petgraph? Like I said, I'm new to petgraph, so maybe I'm missing something obvious, but it just seems strange to me.

TL;DR Why do all methods panic instead of using recoverable errors by returning a Result<...>?

**Comment by starovoid** [comment]
I will start working on this soon and expect the changes will be included in the `0.6.7` release.

**Comment by starovoid** [comment]
## Progress
Implement methods with recoverable errors 
1. [ ] for `Graph` and `StableGraph` #718 :
      - [ ] Add `GraphError` type
      - [ ] `try_add_node`
      - [ ] `try_add_edge`
      - [ ] `try_update_edge`
3. [ ] for `MatrixGraph`:
      - [ ] Add `MatrixError` type
      - [ ] `try_add_node`
      - [ ] `try_add_edge` 
      - [ ] `try_remove_edge`
      - [ ] `try_has_edge`
      - [ ] `get_node_weight` - like `node_weight` but returns `Option<&N>`
      - [ ] `get_node_weight_mut` - like `node_weight_mut` but returns `Option<&mut N>`
      - [ ] `get_edge_weight` - like `edge_weight` but returns `Option<&E>`
      - [ ] `get_edge_weight_mut` - like `edge_weight_mut` but returns `Option<&mut E>`
4. [ ] for `Csr`:
      - [ ] `try_add_edge`
      For other `Csr` methods, the best solution would be to change the overall approach to error handling. 
      Otherwise, to write `try` counterparts, we will need to duplicate almost the entire code of the `Csr` implementation.

**Comment by starovoid** [comment]
@XVilka @TerranWorks 
Hi, I have implemented my vision of checked methods in three PRS, according to the comment above:
- #718 
- #719 
- #720
